File: ios/alier_dev/alier/_LaunchManager.swift
/*
Copyright 2024 Suredesigns Corp.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import Foundation
import UIKit

class LaunchManager {
    private var _launch_apps: [String: String] = [:]
    private var _startup_params: [String: Any?]? = nil
    
    public func getStartupParams() -> [String:Any?] {
        return self._startup_params ?? [:]
    }
    
    public func loadStartupParams(data: URL, eventHandler: EventHandler) {
        var params: [String: Any?] = _startup_params ?? [:]
        _startup_params = params

        setData(data: data, params: &params)
        notifyAwoken(eventHandler: eventHandler, params: params)
    }
    
    public func loadStartupParams(shortcuts: UIApplicationShortcutItem?, eventHandler: EventHandler) {
        var params: [String: Any?] = _startup_params ?? [:]
        _startup_params = params

        if let shortcuts_ = shortcuts {
            setShortcuts(shortcuts: shortcuts_, params: &params)
        }

        notifyAwoken(eventHandler: eventHandler, params: params)
    }
    
    private func notifyAwoken(eventHandler: EventHandler, params: [String: Any?]) {
        try! eventHandler.post(
            category: "lifeCycle",
            message: LifecycleKind.onAwake.toDictionary(params)
        )

    }
    
    private func setData(data: URL, params: inout [String:Any?]) {
        guard let query = data._query(percentEncoded: true) else {
            return
        }
        
        let query_params = query.split(separator: "&")
        if query_params.isEmpty  {
            return
        }
        
        for query_param in query_params {
            let param_pair = query_param.split(separator: "=", maxSplits: 1)
            if param_pair.count == 2 {
                params.updateValue(String(param_pair[1]), forKey: String(param_pair[0]))
            }
        }
    }
    
    private func setShortcuts(shortcuts: UIApplicationShortcutItem, params: inout [String: Any?]) {
        guard let user_info = shortcuts.userInfo else {
            return
        }
        
        let keys = [
            "shortcut",
            "shortcut1",
            "shortcut2",
            "shortcut3",
            "shortcut4",
        ]
        
        for key in keys {
            if let value = user_info[key] as? String {
                params[key] = value
            }
        }
    }
    
        
    public func registerLaunchApp(name: String, uri: String) {
        _launch_apps.updateValue(uri, forKey: name)
    }
    
    public func launchOtherApp(action: String, params: String) {
        // https://developer.apple.com/library/archive/featuredarticles/iPhoneURLScheme_Reference/Introduction/Introduction.html
        switch action {
        case "alarm" : do {
        }
        case "set_alarm" : do {
        }
        case "timer" : do {
        }
        case "set_timer" : do {
        }
        case "calendar" : do {
            let url = URL(string: "calshow://" + setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        case "camera" : do {
        }
        case "camera_capture" : do {
        }
        case "video" : do {
            let url = URL(string: "videos://" + setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        case "video_capture" : do {
        }
        case "mail" : do {
            let url = URL(string: "mailto://" + setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        case "map" : do {
            let url = URL(string: "maps://" + setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        case "memo" : do {
            let url = URL(string: "mobilenotes://showNote" + setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        case "tel" : do {
            // Does not work because the phone app does not exist on the simulator
            let url = URL(string: "tel://" + setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        case "facetime" : do {
            // iOS only
            let url = URL(string: "facetime://" + setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        case "facetime_audio" : do {
            // iOS only
            let url = URL(string: "facetime-audio://" + setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        case "search" : do {
        }
        case "settings" : do {
            // Deprecated in iOS5.1
            let url = URL(string: "prefs://" + setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        case "sms" : do {
            let url = URL(string: "sms://" + setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        case "mms" : do {
            let url = URL(string: "sms://" + setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        case "browser" : do {
            // https cannot be used
            let url = URL(string: setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        case "appstore" : do {
            let url = URL(string: "itms-apps://" + setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        case "itunes" : do {
            let url = URL(string: "http://phobos.apple.com/" + setLaunchParam(action: action, params: params))
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url!, options: [:])
            } else {
                UIApplication.shared.openURL(url!)
            }
        }
        default : do {
            if _launch_apps.keys.contains(action) {
                let scheme = _launch_apps[action]
                if let url = URL(string: scheme!) {
                    if #available(iOS 10.0, *) {
                        UIApplication.shared.open(url, options: [:])
                    } else {
                        UIApplication.shared.openURL(url)
                    }
                } else {
                    AlierLog.w(id: 0, message: scheme! + " not installed.")
                }
            } else {
                AlierLog.w(id: 0, message: "No action registered : " + action)
            }
        }
        }
    }
    
    private func setLaunchParam(action: String, params: String) -> String {
        var res = ""
        let data: Data =  params.data(using: String.Encoding.utf8)!
        var jsonDict = Dictionary<String, Any>()
        do {
            jsonDict = try JSONSerialization.jsonObject(with: data) as! Dictionary<String, Any>
            // alarm
            // set_alarm
            // timer
            // set_timer
            // calendar
            // camera
            // camera_capture
            // video
            // video_capture
            // mail
            if (jsonDict.keys.contains("to")) {
                res += cnvAny2String(jsonDict["to"])
            }
            if (jsonDict.keys.contains("subject")) {
                if (!res.contains("?")) { res += "?" }
                res += "subject=" + cnvAny2String(jsonDict["subject"]) + "&"
            }
            // map
            if (jsonDict.keys.contains("latitude") && jsonDict.keys.contains("longitude")) {
                res += "ll=" + cnvAny2String(jsonDict["latitude"]) + "," + cnvAny2String(jsonDict["longitude"])
            }
            if (jsonDict.keys.contains("zoom")) {
                if (!res.contains("?")) { res += "?" }
                res += "z=" + cnvAny2String(jsonDict["zoom"]) + "&"
            }
            if (jsonDict.keys.contains("label_lat") && jsonDict.keys.contains("label_lng")) {
                if (!res.contains("?")) { res += "?" }
                res += "sll=" + cnvAny2String(jsonDict["label_lat"]) + "," + cnvAny2String(jsonDict["label_lng"]) + "&"
            }
            if (jsonDict.keys.contains("house") && jsonDict.keys.contains("street") && jsonDict.keys.contains("address")) {
                if (!res.contains("?")) { res += "?" }
                res += "address=" + cnvAny2String(jsonDict["house"]) + "," + cnvAny2String(jsonDict["street"]) + "," + cnvAny2String(jsonDict["address"]) + "&"
            }
            if (jsonDict.keys.contains("type")) {
                if (!res.contains("?")) { res += "?" }
                res += "t=" + cnvAny2String(jsonDict["type"]) + "&"
            }
            // memo
            // tel
            if (jsonDict.keys.contains("number")) {
                res += cnvAny2String(jsonDict["number"])
            }
            // search
            // settings
            // sms/mms
            if (jsonDict.keys.contains("text")) {
                if (!res.contains("?")) { res += "?" }
                res += "body=" + cnvAny2String(jsonDict["text"]) + "&"
            }
            // browser
            if (jsonDict.keys.contains("uri")) {
                res += cnvAny2String(jsonDict["uri"])
            }
            // appstore
            // itunes
            if (jsonDict.keys.contains("store")) {
                res += cnvAny2String(jsonDict["store"])
            }
        } catch {
            AlierLog.e(id: 0, message: error.localizedDescription)
        }
        
        if res.last == "&" {
            res = String(res.dropLast())
        }
        return res
    }
    
    private func cnvAny2String(_ data: Any?) -> String {
        switch (data) {
        case let intValue as Int:
            return String(intValue)
        case let stringValue as String:
            return stringValue
        case let dateValue as Date:
            let f = DateFormatter()
            f.dateFormat = "yyyy/mm/dd"
            return f.string(from: dateValue)
        default:
            return "?"
        }
    }
}

File: android/alier_dev/alier/src/main/java/com/suredesigns/alier/_LaunchManager.kt
/*
Copyright 2024 Suredesigns Corp.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package com.suredesigns.alier

import android.app.Activity
import android.app.SearchManager
import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.AlarmClock
import android.provider.CalendarContract
import android.provider.MediaStore
import android.provider.Settings
import com.google.android.gms.actions.NoteIntents
import org.json.JSONObject
import java.lang.Exception
import androidx.core.net.toUri

class _LaunchManager {
    private val _launch_apps = mutableMapOf<String, String>()
    private var _startup_params: Map<String, Any?>? = null
    private var _uri_host: String = ""

    fun getStartupParams(): Map<String, Any?> {
        return _startup_params ?: mapOf()
    }

    fun getUriHost(): String {
        return _uri_host
    }

    fun loadStartupParams(intent: Intent, eventHandler: EventHandler) {
        val params =  mutableMapOf<String, Any?>()

        // Get startup parameters when launched by URL and other apps and push notification
        val data = intent.data
        if (data != null) {
            setData(data, params)
        }

        val extras = intent.extras
        if (extras != null) {
            setShortcuts(extras, params)
            setWidgets(extras, params)
        }

        val startup_params = params.toMap()
        _startup_params = startup_params

        notifyAwoken(eventHandler, startup_params)
    }

    private fun notifyAwoken(event: EventHandler, params: Map<String, Any?>) {
        event.post(
            category = "lifeCycle",
            message = BaseMainActivity.LifecycleKind.onAwake.toMap(params)
        )
    }

    private fun setData(data: Uri, params: MutableMap<String, Any?>) {
        val host = data.host
        if (host != null) {
            _uri_host = host
        }
        for (key in data.queryParameterNames) {
            //  Use the first value if there are duplicated values for the given key.
            params[key] = data.getQueryParameter(key)
        }
    }

    // Launch in shortcut
    private fun setShortcuts(extras: Bundle, params: MutableMap<String, Any?>) {
        val keys = arrayOf(
            "shortcut",
            "shortcut1",
            "shortcut2",
            "shortcut3",
            "shortcut4",
            )

        for (key in keys) {
            val shortcut = extras.getString(key)
            if (shortcut.isNullOrEmpty()) { continue }

            params[key] = shortcut
        }
    }

    // Launch in widget
    private fun setWidgets(extras: Bundle, params: MutableMap<String, Any?>) {
        val keys = arrayOf(
            "widget",
            "widget1",
            "widget2",
            "widget3",
            "widget4",
            "widget5",
        )

        for (key in keys) {
            val widget = extras.getString(key)
            if (widget.isNullOrEmpty()) { continue }

            params[key] = widget
        }
    }

    fun registerLaunchApp(name: String, uri: String) {
        _launch_apps[name] = uri
    }

    fun launchOtherApp(activity: Activity, action: String, params: String) {
        val other_app_intent = Intent()
        // https://developer.android.com/guide/components/intents-common?hl=ja
        when (action) {
            "alarm" -> {
                other_app_intent.setAction(AlarmClock.ACTION_SHOW_ALARMS)
            }
            "timer" -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    other_app_intent.setAction(AlarmClock.ACTION_SHOW_TIMERS)
                } else {
                    other_app_intent.setAction("android.intent.action.SHOW_TIMERS")
                }
            }
            "set_alarm" -> {
                other_app_intent.setAction(AlarmClock.ACTION_SET_ALARM).apply {
                    setLaunchParam(other_app_intent, action, params)
                }
            }
            "set_timer" -> {
                other_app_intent.setAction(AlarmClock.ACTION_SET_TIMER).apply {
                    setLaunchParam(other_app_intent, action, params)
                }
            }
            "calendar" -> {
                other_app_intent.setData(CalendarContract.Events.CONTENT_URI)
                other_app_intent.setAction(Intent.ACTION_INSERT).apply {
                    setLaunchParam(other_app_intent, action, params)
                }
            }
            "camera" -> {
                other_app_intent.setAction(MediaStore.INTENT_ACTION_STILL_IMAGE_CAMERA)
            }
            "camera_capture" -> {
                other_app_intent.setAction(MediaStore.ACTION_IMAGE_CAPTURE).apply {
                    setLaunchParam(other_app_intent, action, params)
                }
            }
            "video" -> {
                other_app_intent.setAction(MediaStore.INTENT_ACTION_VIDEO_CAMERA)
            }
            "video_capture" -> {
                other_app_intent.setAction(MediaStore.ACTION_VIDEO_CAPTURE).apply {
                    setLaunchParam(other_app_intent, action, params)
                }
            }
            "mail" -> {
                other_app_intent.setAction(Intent.ACTION_SEND).apply {
                    other_app_intent.setType("*/*")
                    setLaunchParam(other_app_intent, action, params)
                }
            }
            "map" -> {
                var data = "geo:"
                other_app_intent.setAction(Intent.ACTION_VIEW).apply {
                    data += setLaunchParam(other_app_intent, action, params)
                }
                other_app_intent.setData(data.toUri())
            }
            "memo" -> {
                other_app_intent.setAction(NoteIntents.ACTION_CREATE_NOTE).apply {
                    setLaunchParam(other_app_intent, action, params)
                }
            }
            "tel" -> {
                var data = "tel:"
                other_app_intent.setAction(Intent.ACTION_DIAL).apply {
                    data += setLaunchParam(other_app_intent, action, params)
                }
                other_app_intent.setData(data.toUri())
            }
            "facetime" -> {

            }
            "facetime_audio" -> {

            }
            "search" -> {
                other_app_intent.setAction(Intent.ACTION_WEB_SEARCH).apply {
                    setLaunchParam(other_app_intent, action, params)
                }
            }
            "settings" -> {
                other_app_intent.setAction(Settings.ACTION_WIFI_SETTINGS)
            }
            "sms", "mms" -> {
                var data = "sms:"
                other_app_intent.setAction(Intent.ACTION_SENDTO).apply {
                    other_app_intent.setType("*/*")
                    data += setLaunchParam(other_app_intent, action, params)
                }
                other_app_intent.setData(data.toUri())
            }
            "browser" -> {
                var data = ""
                other_app_intent.setAction(Intent.ACTION_VIEW).apply {
                    data += setLaunchParam(other_app_intent, action, params)
                }
                other_app_intent.setData(data.toUri())
            }
            "appstore" -> {

            }
            "itunes" -> {

            }
            else -> {
                val data = _launch_apps[action]
                if (data != null) {
                    other_app_intent.setAction(Intent.ACTION_VIEW)
                        .setData(data.toUri())
                        .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                        other_app_intent
                            .addFlags(Intent.FLAG_ACTIVITY_REQUIRE_DEFAULT)
                            .addFlags(Intent.FLAG_ACTIVITY_REQUIRE_NON_BROWSER)
                    }

                    try {
                        activity.startActivity(other_app_intent)
                    } catch (e: ActivityNotFoundException) {
                        AlierLog.w(0, "$data not installed.")
                    }
                } else {
                    AlierLog.w(0, "No action registered : $action")
                }
                return
            }
        }

        try {
            activity.startActivity(other_app_intent)
        } catch (e: Exception) {
            AlierLog.w(0, "Failed to start $action app.")
        }
    }

    private fun setLaunchParam(intent: Intent, action: String, params: String) : String {
        var data = ""
        if (params.isNotEmpty() && params != "\"\"") {
            val json = JSONObject(params)
            // set_alarm
            if (!json.isNull("message")) {
                intent.putExtra(AlarmClock.EXTRA_MESSAGE, json.getString("message"))
            }
            if (!json.isNull("days")) {
                // Integer in calendar class
                intent.putExtra(AlarmClock.EXTRA_DAYS, json.getInt("days"))
            }
            if (!json.isNull("hour")) {
                intent.putExtra(AlarmClock.EXTRA_HOUR, json.getString("hour"))
            }
            if (!json.isNull("minutes")) {
                intent.putExtra(AlarmClock.EXTRA_MINUTES, json.getString("minutes"))
            }
            if (!json.isNull("ringtone")) {
                intent.putExtra(AlarmClock.EXTRA_RINGTONE, json.getString("ringtone"))
            }
            if (!json.isNull("vibrate")) {
                intent.putExtra(AlarmClock.EXTRA_VIBRATE, json.getBoolean("vibrate"))
            }
            if (!json.isNull("skipUI")) {
                intent.putExtra(AlarmClock.EXTRA_SKIP_UI, json.getBoolean("skipUI"))
            }
            // set_timer
            if (!json.isNull("length")) {
                // seconds
                intent.putExtra(AlarmClock.EXTRA_LENGTH, json.getInt("length"))
            }
            if (!json.isNull("skipUI")) {
                intent.putExtra(AlarmClock.EXTRA_SKIP_UI, json.getBoolean("skipUI"))
            }
            // calendar
            if (!json.isNull("title")) {
                intent.putExtra(CalendarContract.Events.TITLE, json.getString("title"))
            }
            if (!json.isNull("description")) {
                intent.putExtra(CalendarContract.Events.DESCRIPTION, json.getString("description"))
            }
            if (!json.isNull("location")) {
                intent.putExtra(CalendarContract.Events.EVENT_LOCATION, json.getString("location"))
            }
            if (!json.isNull("all_day")) {
                intent.putExtra(CalendarContract.EXTRA_EVENT_ALL_DAY, json.getBoolean("all_day"))
            }
            if (!json.isNull("begin_time")) {
                // milliseconds
                intent.putExtra(CalendarContract.EXTRA_EVENT_BEGIN_TIME, json.getInt("begin_time"))
            }
            if (!json.isNull("end_time")) {
                intent.putExtra(CalendarContract.EXTRA_EVENT_END_TIME, json.getInt("end_time"))
            }
            // camera_capture, video_capture
            if (!json.isNull("output")) {
                intent.putExtra(MediaStore.EXTRA_OUTPUT, json.getString("output"))
            }
            // mail
            if (!json.isNull("to")) {
                intent.putExtra(Intent.EXTRA_EMAIL, json.getString("to"))
            }
            if (!json.isNull("cc")) {
                intent.putExtra(Intent.EXTRA_CC, json.getString("cc"))
            }
            if (!json.isNull("bcc")) {
                intent.putExtra(Intent.EXTRA_BCC, json.getString("bcc"))
            }
            if (!json.isNull("subject")) {
                intent.putExtra(Intent.EXTRA_SUBJECT, json.getString("subject"))
            }
            if (!json.isNull("text")) {
                intent.putExtra(Intent.EXTRA_TEXT, json.getString("text"))
            }
            if (!json.isNull("stream")) {
                intent.putExtra(Intent.EXTRA_STREAM, json.getString("stream"))
            }
            // map
            if (!json.isNull("latitude") && !json.isNull("longitude")) {
                data += json.getString("latitude") + "," + json.getString("longitude")
            }
            if (!json.isNull("zoom")) {
                data += "?z=" + json.getInt("zoom").toString()
            }
            if (!json.isNull("label") && !json.isNull("label_lat") && !json.isNull("label_lng")) {
                val label = json.getString("label")
                val label_lat = json.getString("label_lat")
                val label_lng = json.getString("label_lng")
                data += if (!json.isNull("zoom")) "&" else "?"
                data += "q=${label_lat},${label_lng}(${label})"
            } else if (!json.isNull("house") && !json.isNull("street") && !json.isNull("address")) {
                val house = json.getString("house")
                val street = json.getString("street")
                val address = json.getString("address")
                data += if (!json.isNull("zoom")) "&" else "?"
                data += "q=${house}+${street}+${address}"
            }
            // memo
            if (!json.isNull("title")) {
                intent.putExtra(NoteIntents.EXTRA_NAME, json.getString("title"))
            }
            if (!json.isNull("text")) {
                intent.putExtra(NoteIntents.EXTRA_TEXT, json.getString("text"))
            }
            // tel
            if (!json.isNull("number")) {
                data += json.getString("number")
            }
            // search
            if (!json.isNull("query")) {
                intent.putExtra(SearchManager.QUERY, json.getString("query"))
            }
            // sms/mms
            if (!json.isNull("subject")) {
                intent.putExtra("subject", json.getString("subject"))
            }
            if (!json.isNull("text")) {
                intent.putExtra("sms_body", json.getString("text"))
            }
            if (!json.isNull("stream")) {
                intent.putExtra(Intent.EXTRA_STREAM, json.getString("stream"))
            }
            // browser
            if (!json.isNull("uri")) {
                data += json.getString("uri")
            }
        }
        return data
    }
}